File: 2 - Sorting and Order Statistics/2.2 - Quick Sort/Code/randomized_quicksort.py
"""
randomized_quicksort.py

"""

from random import randint
from .quicksort import Quicksort


class RandomizedQuicksort(Quicksort):
    """Randomized Quicksort"""

    def randomized_partition(self, p, r):
        """Implements 'partition' from Quicksort with randomized pivot element"""
        i = randint(p, r)
        self.array[r], self.array[i] = self.array[i], self.array[r]
        return self.partition(p, r)

    def randomized_quicksort(self, p, r):
        """Quicksort with the 'randomized_partition' method"""
        if p < r:
            q = self.randomized_partition(p, r)
            self.randomized_quicksort(p, q - 1)
            self.randomized_quicksort(q + 1, r)

File: 2 - Sorting and Order Statistics/2.2 - Quick Sort/Code/RandomizedQuicksort.js
/**
 * RandomizedQuicksort.js
 *
 */

const Quicksort = require("./Quicksort");

class RandomizedQuicksort extends Quicksort {
    constructor(array) {
        super(array);
    }

    randomizedPartition(p, r) {
        let x = this.array[r];
        let i = p - 1;

        for (let j = p; j < r; j++) {
            if (this.array[j] <= x) {
                i += 1;
                let temp = this.array[i];
                this.array[i] = this.array[j];
                this.array[j] = temp;
            }
        }

        let temp = this.array[i + 1];
        this.array[i + 1] = this.array[r];
        this.array[r] = temp;

        return i + 1;
    }

    randomizedQuicksort(p, r) {
        if (p < r) {
            let q = this.partition(p, r);
            this.randomizedQuicksort(p, q - 1);
            this.randomizedQuicksort(q + 1, r);
        }
    }
}

module.exports = RandomizedQuicksort;
